ee/server/service/mdm.go

```diff
@@ -742,7 +745,104 @@ func (svc *Service) MDMAppleMatchPreassignment(ctx context.Context, externalHost
 		return ctxerr.Wrap(ctx, err)
 	}
 
-	// TODO(mna): TBD the exact signature, must return the list of profiles for matching
-	//profs, err := svc.profileMatcher.RetrieveProfiles(ctx, externalHostIdentifier)
+	profs, err := svc.profileMatcher.RetrieveProfiles(ctx, externalHostIdentifier)
+	if err != nil {
+		return err
+	}
+	if len(profs.Profiles) == 0 || profs.HostUUID == "" {
+		return nil // nothing to do
+	}
+
+	// load the host and ensure it is enrolled in Fleet MDM
+	host, err := svc.ds.HostByIdentifier(ctx, profs.HostUUID)
+	if err != nil {
+		return err // will return a not found error if host does not exist
+	}
+
+	hostMDM, err := svc.ds.GetHostMDM(ctx, host.ID)
+	if err != nil || !hostMDM.IsFleetEnrolled() {
+		if err == nil || fleet.IsNotFound(err) {
+			err = errors.New("host is not enrolled in Fleet MDM")
+			return ctxerr.Wrap(ctx, &fleet.BadRequestError{
+				Message:     err.Error(),
+				InternalErr: err,
+			})
+		}
+		return err
+	}
+
+	// Collect the profiles' hashes and look for a team with exactly that set.
+	// Also collect the profiles' groups in case we need to create a new team,
+	// and the list of raw profiles bytes.
+	hashes, groups, rawProfiles := make([]string, 0, len(profs.Profiles)),
+		make([]string, 0, len(profs.Profiles)),
+		make([][]byte, 0, len(profs.Profiles))
+	for _, prof := range profs.Profiles {
+		hashes = append(hashes, prof.HexMD5Hash)
+		rawProfiles = append(rawProfiles, prof.Profile)
+		if prof.Group != "" {
+			groups = append(groups, prof.Group)
+		}
+	}
+
+	// find a team with exactly that set of profiles
+	teamIDs, err := svc.ds.MatchMDMAppleConfigProfiles(ctx, hashes)
+	if err != nil {
+		return err
+	}
+
+	var targetTeamID uint
+	if len(teamIDs) > 0 {
+		// if the host is already in one of those valid teams, nothing to do.
+		if host.TeamID != nil {
+			for _, tmID := range teamIDs {
+				if *host.TeamID == tmID {
+					return nil
+				}
+			}
+		}
+		// else assign the host to the first valid team
+		targetTeamID = teamIDs[0]
+
+	} else {
+		// create a new team with this set of profiles
+		dedupeGroups := make(map[string]struct{}, len(groups))
+		for _, group := range groups {
+			dedupeGroups[group] = struct{}{}
+		}
+		groups = groups[:0]
+		for group := range dedupeGroups {
+			groups = append(groups, group)
+		}
+		sort.Strings(groups)
+
+		if len(groups) == 0 {
+			groups = []string{"default"}
+		}
+
+		// creating via the service call so that it properly assigns the agent
+		// options and creates audit activities, etc.
+		teamName := fmt.Sprintf("%s (%s)", strings.Join(groups, " - "), time.Now().UTC().Format("2006-01-02:15:04:05"))
```

Comment:
The new "group" concept in relation to "teams" takes some time to unpack. Is there a good place where we can leave some breadcrumbs for our future selves?

Reply:
:+1: yes, looking at it I think all that group cleanup and team name building could be extracted to e.g. a `teamNameFromGroups` function with a comment there explaining what role the groups play (and the function name already setting some context), does that make sense to you?